parquet-column/src/test/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,412 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.column.values.bloom;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.IntBuffer;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.parquet.Preconditions;
+import org.apache.parquet.bytes.*;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.io.ParquetEncodingException;
+import org.apache.parquet.io.api.Binary;
+
+/**
+ * Bloom Filter is a compact structure to indicate whether an item is not in set or probably
+ * in set. Bloom class is underlying class of Bloom Filter which stores a bit set represents
+ * elements set, hash strategy and bloom filter algorithm.
+ *
+ * Bloom Filter algorithm is implemented using block Bloom filters from Putze et al.'s "Cache-,
+ * Hash- and Space-Efficient Bloom Filters". The basic idea is to hash the item to a tiny Bloom
+ * Filter which size fit a single cache line or smaller. This implementation sets 8 bits in
+ * each tiny Bloom Filter. Tiny bloom filter are 32 bytes to take advantage of 32-bytes SIMD
+ * instruction.
+ */
+
+public class Bloom {
+  // Hash strategy available for bloom filter.
+  public enum HASH {
+    MURMUR3_X64_128,
+  }
+
+  // Bloom filter algorithm.
+  public enum ALGORITHM {
+    BLOCK,
+  }
+
+  /**
+   * Default false positive probability value use to calculate optimal number of bits
+   * used by bloom filter.
+   */
+  public final double DEFAULT_FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  public static final int HEADER_SIZE = 12;
+
+  // Bytes in a bucket.
+  public static final int BYTES_PER_BUCKET = 32;
+
+  // Hash strategy used in this bloom filter.
+  public final HASH hash;
+
+  // Algorithm applied of this bloom filter.
+  public final ALGORITHM algorithm;
+
+  // The underlying byte array for bloom filter bitset.
+  private byte[] bitset;
```

Comment:
Why not just store int[] bitset?

Reply:
Since when serializing and deserializing you will need to additional considering about the byte order and converting byte array to int array which should have some overhead.

Reply:
This doubles the memory requirement, and I don't think it is worth it -- doubly so since you can change the endianness with the subtraction bit below.

Reply:
The intBuffer just use bitset as backup array.